cpp/src/io/text/bgzip_data_chunk_source.cu

```diff
@@ -0,0 +1,451 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "io/comp/nvcomp_adapter.hpp"
+
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/io/text/data_chunk_source_factories.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/host_vector.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/system/cuda/experimental/pinned_allocator.h>
+#include <thrust/transform.h>
+
+#include <fstream>
+#include <limits>
+
+namespace cudf::io::text {
+
+namespace {
+
+class device_uvector_data_chunk : public device_data_chunk {
+ public:
+  device_uvector_data_chunk(rmm::device_uvector<char>&& data) : _data(std::move(data)) {}
+
+  [[nodiscard]] char const* data() const override { return _data.data(); }
+  [[nodiscard]] std::size_t size() const override { return _data.size(); }
+  operator device_span<char const>() const override { return _data; }
+
+ private:
+  rmm::device_uvector<char> _data;
+};
+
+struct bgzip_nvcomp_transform_functor {
+  uint8_t const* compressed_ptr;
+  uint8_t* decompressed_ptr;
+
+  __device__ thrust::tuple<device_span<const uint8_t>, device_span<uint8_t>> operator()(
+    thrust::tuple<std::size_t, std::size_t, std::size_t, std::size_t> t)
+  {
+    auto const compressed_begin   = thrust::get<0>(t);
+    auto const compressed_end     = thrust::get<1>(t);
+    auto const decompressed_begin = thrust::get<2>(t);
+    auto const decompressed_end   = thrust::get<3>(t);
+    return thrust::make_tuple(device_span<const uint8_t>{compressed_ptr + compressed_begin,
+                                                         compressed_end - compressed_begin},
+                              device_span<uint8_t>{decompressed_ptr + decompressed_begin,
+                                                   decompressed_end - decompressed_begin});
+  }
+};
+
+class bgzip_data_chunk_reader : public data_chunk_reader {
+ private:
+  /*
+
+
+     Parsing code
+
+
+  */
+
+  template <typename IntType>
+  static bool is_little_endian()
+  {
+    IntType i = 0x0100;
+    std::array<char, sizeof(i)> bytes;
+    std::memcpy(&bytes[0], &i, sizeof(i));
+    return bytes[0] == 0;
+  }
```

Comment:
Do we really need this? I would assume that every platform we support is little-endian.